### ISSUE TITLE
Windows support

### ISSUE THREAD

**Issue by fschulze** [lektor/lektor#121]
Gathering info on the state of Windows support here. The final goal is an installer.

From the top of my head the list of things I did to get the GUI running:
- install git for Windows (https://git-for-windows.github.io) this includes bash
- install nodejs for Windows (https://nodejs.org/en/download/) this includes npm
- install Visual Studio Express Community Edition (https://www.visualstudio.com/products/visual-studio-express-vs) it seems like this includes git and nodejs for windows, I saw it greyed out in the custom install options
- with bash in your checkout `cd lektor/admin`, `npm install .`, `cd static`, `../node_modules/.bin/webpack` (the git bash has no `make`, otherwise `make build-js` should work)
- with bash in your checkout `cd gui`, `npm install --msvs_version=2015 .` (might need to be updated for newer Visual Studio releases, I also had to remove `node_modules` which was generated before Visual Studio was installed)
- create a virtualenv in a separate directory (there seem to be issues if it's in the checkout)
- activate the virtualenv
- install Lektor with `pip install -e [path to checkout]`
- run the gui with `[path to checkout]/gui/bin/run`

All the basics I tried worked from there.

**Comment by fschulze** [comment]
Possible solution for packaging:
- https://discuss.atom.io/t/creating-windows-installer-on-mac/22004/2

**Comment by datnamer** [comment]
How about a conda package

**Comment by karelv** [comment]
or choco https://chocolatey.org/

**Comment by joelgoop** [comment]
I managed to package the electron GUI into a Windows .exe using `electron-packager` and built it into an installer with `electron-builder`. However, there are some issues:
- It still requires the `lektor` command itself to be installed separately
- The total size of the GUI package is around 100 MB which seems unreasonably large for such a simple app
- It is very slow to start on Windows (30+ seconds on my machine)
- Killing process trees spawned as subprocesses doesn't work on Windows, which means that "Close Project" does nothing, see quick fix below

Is there a reason for not building a GUI directly in Python that can call `lektor` directly?

### Details

#### Quick fix for process kill

To make "Close Project" work, I had to change the `LektorServer.shutdown` (in gui/static/js/lektorInterop.jsx) function to:

```
  shutdown() {
    if (process.platform == 'win32') {
      childProcess.spawnSync('taskkill /T /F /PID ' + this._child.pid)
    }else{
      this._child.kill('SIGHUP');  
    } 
  }
```

(Based on: http://stackoverflow.com/questions/32705857/cant-kill-child-process-on-windows)

#### Packaging

I tried making the packaging work under Windows, but it was much easier to do under Ubuntu and the result seemed to work fine on Windows. For packaging, I did `electron-packager [path-to-gui] LektorGUI --platform=win32 --arch=ia32 --version=0.34.3 --prune --icon=icon.ico`.

#### Building the installer

For building the installer I installed `wine` and `nsis`:

```
sudo add-apt-repository ppa:ubuntu-wine/ppa
sudo apt-get install wine nsis
```

I added configuration to package.json:

```
  "builder": {
    "win" : {
      "title" : "LektorGUI",
      "version" : "1.0.0",
      "icon": "icon.ico",
      "publisher": "Lektor"
    }
  }
```

Then I ran `electron-builder [path-to-package-folder] --platform=win --out=[output-path]`

**Comment by fschulze** [comment]
Since the GUI is basically just opening the Browser, a simple PySide or even Tk GUI could work and would simplify things quite a bit, wouldn't it? I guess the initial idea was to have the admin interface inside the elektron app?

**Comment by joelgoop** [comment]
Yes, I was also thinking something simple like TkInter, which should make it easy to build a minimal GUI to open projects and control `lektor server`. Packaging the whole program for Windows is a different story, though. I played around a little with PyInstaller, but it didn't seem to work with lektor's dependency management (haven't looked into any details). Is there a way to create an installer for lektor on Windows which also allows it to install its own packages? But that has to be solved anyway, since the main lektor app has to be packaged independently of how the GUI is built.

**Comment by datnamer** [comment]
Yes, conda should be able to package lektor.

**Comment by meatballs** [comment]
Given that the [BeeWare project](http://pybee.org/project) is using lektor for its website, it might be good to use their toolkit for a cross-platform gui. @freakboy3742

**Comment by freakboy3742** [comment]
I wouldn't object to the reciprocity, but it certainly isn't expected. In this specific case, Beeware (and Toga specifically) is _way_ too early stage to be used seriously. However, soon, maybe...

**Comment by goanpeca** [comment]
I would also like to make a conda package for this (with conda-forge most likely). I can make the package :-) (most of the conda pacakges needed are there already)

But according to...

![image](https://cloud.githubusercontent.com/assets/3627835/17874089/17bb0fac-688f-11e6-9978-2e28a4750191.png)

This might be problematic, what are the implications of `Lektor actively manages virtualenvs for plugin installations in very specific ways` @mitsuhiko?

**Comment by goanpeca** [comment]
Conda packages are on the works [here](https://github.com/conda-forge/staged-recipes/pull/1350)

This should enable lektor (including nodejs/npm) on all major platforms:

`conda install lektor -c conda-forge`

**Comment by goanpeca** [comment]
Conda packages for lektor are ready :-)
